### PR TITLE
[Ledger]: Expose transaction fields to smart contracts

### LINKED ISSUE #1266: Expose transaction fields to smart contracts
Enable smart contracts to access the amount of FET sent along a transaction

### PR DESCRIPTION
**Pull request by pbukva**
Resolves #1266

### CODE REVIEW

File: libs/ledger/src/chaincode/contract.cpp [hunk 1]
```diff
@@ -31,13 +31,14 @@ namespace ledger {
  * @param tx The reference to the originating transaction
  * @return The corresponding status result for the operation
  */
-Contract::Result Contract::DispatchInitialise(Address const &owner)
+Contract::Result Contract::DispatchInitialise(Address const &owner, Transaction const &tx,
+                                              Transaction::BlockIndex block_index)
```
Comment (bipll) [inline libs/ledger/src/chaincode/contract.cpp]: `BlockIndex` is aliased in hpp.

File: libs/ledger/tests/chaincode/smart_contract_tests.cpp [hunk 2]
```diff
@@ -567,4 +570,357 @@ TEST_F(SmartContractTests, CheckShardedStateSetWithAddressAsName)
   VerifyQuery("query_foo", int32_t{20}, request);
 }
 
+bool Replace(std::string &orig, std::string const &what, std::string const &with)
```
Comment (bipll) [inline libs/ledger/tests/chaincode/smart_contract_tests.cpp]: This probably can be made a library function, if there does not exist one already.

File: libs/vm_modules/include/vm_modules/ledger/transfer.hpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#include "vm/address.hpp"
+#include "vm_modules/ledger/forward_decl.hpp"
+
+namespace fetch {
+namespace vm_modules {
+namespace ledger {
+
+class Transfer : public vm::Object
+{
+  AddressPtr        address_{};
```
Comment (bipll) [inline libs/vm_modules/include/vm_modules/ledger/transfer.hpp]: `{}` probably not needed here.

**Review by bipll — DISMISSED** [review]

File: libs/vm/include/vm/address.hpp [hunk 4]
```diff
@@ -129,6 +128,11 @@ class Address : public Object
     return true;
   }
 
+  std::size_t GetHashCode() override
+  {
+    return std::hash<ledger::Address>{}(address_) /* ^ std::hash<bool>{}(signed_tx_)*/;
```
Comment (ejfitzgerald) [inline libs/vm/include/vm/address.hpp]: Remove comment?

File: libs/vm_modules/CMakeLists.txt [hunk 5]
```diff
@@ -19,7 +19,8 @@ target_link_libraries(fetch-vm-modules
                       PUBLIC fetch-vm
                              fetch-core
                              fetch-math
-                             fetch-ml)
+                             fetch-ml
+                             fetch-ledger)
```
Comment (ejfitzgerald) [inline libs/vm_modules/CMakeLists.txt:23]: Does this introduce a circular dependency?

File: libs/vm_modules/include/vm_modules/ledger/transaction.hpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  AddressPtr                                  from_;
+  TransfersPtr                                transfers_;
+  AddressPtr                                  contract_digest_;
+  AddressPtr                                  contract_address_;
+  StringPtr                                   action_;
+  AddressesPtr                                signatories_;
+  // BitVectorPtr shard_mask_; //TODO(pb): Implement in the next stage
+
+public:
+  Transaction() = delete;
```
Comment (ejfitzgerald) [inline libs/vm_modules/include/vm_modules/ledger/transaction.hpp]: Strictly not necessary if you have already defined a constructor

File: libs/vm_modules/include/vm_modules/ledger/transfer.hpp [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+#include "vm/address.hpp"
+#include "vm_modules/ledger/forward_decl.hpp"
+
+namespace fetch {
+namespace vm_modules {
+namespace ledger {
+
+class Transfer : public vm::Object
+{
+  AddressPtr        address_;
```
Comment (ejfitzgerald) [inline libs/vm_modules/include/vm_modules/ledger/transfer.hpp]: Does this class need to be this way around?

File: libs/vm_modules/include/vm_modules/ledger/transaction.hpp [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  Transaction(vm::VM *vm, vm::TypeId type_id, fetch::ledger::Transaction const &tx);
+  ~Transaction() override = default;
+
+  static void Bind(vm::Module &module);
+  // static vm::Ptr<Transaction> Constructor(vm::VM *vm, vm::TypeId type_id,
```
Comment (ejfitzgerald) [inline libs/vm_modules/include/vm_modules/ledger/transaction.hpp]: Remove?

File: libs/vm_modules/include/vm_modules/ledger/transaction.hpp [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+{
+  std::shared_ptr<fetch::ledger::Transaction> tx_;
+  DigestPtr                                   digest_;
+  AddressPtr                                  from_;
+  TransfersPtr                                transfers_;
+  AddressPtr                                  contract_digest_;
+  AddressPtr                                  contract_address_;
+  StringPtr                                   action_;
+  AddressesPtr                                signatories_;
+  // BitVectorPtr shard_mask_; //TODO(pb): Implement in the next stage
```
Comment (ejfitzgerald) [inline libs/vm_modules/include/vm_modules/ledger/transaction.hpp]: Can you raise a Jira issue for this and probably remove?

**Review by ejfitzgerald — CHANGES_REQUESTED** [review]

File: libs/vm_modules/CMakeLists.txt (hunk 5 above)
Comment (pbukva) [reply libs/vm_modules/CMakeLists.txt:23]: It has been agreed that this will be sorted out in upcoming change for bazel integration.

**Review by ejfitzgerald — APPROVED** [review]

**Pull request merged**